### DISCUSSION TITLE
Adding images in a canvas layout causes weird behaviour

### DISCUSSION THREAD

**Discussion by mikekok** [fabricjs/fabric.js#7641] [Q&A]
I initially wanted to open an issue about this but I thought I'd post it here first, in case I'm doing something wrong.

So, I'm trying to add multiple images to the canvas in a layout form (kind of like a grid) using a for loop.
The problem is, the first time the images are added to the canvas, the layout is not respected.
Some images are completely skipped and some others are added at the wrong position. But **once I call the same function again**, everything works perfectly fine.

To reproduce this issue, click the "Load Images" button once, have a look at the results and then click it once again in order to get the desired result. If you want to see it again, rerunning the fiddle using the "Run" button on top isn't enough, you'll have to reload the page (which is pretty crazy to me as I can't really figure out what may be causing that).

[jsfiddle](https://jsfiddle.net/79oyvdga/3/)

**Comment by asturur** [comment]
hi @mikekok this is the classic js callback/async issue.
You are doing a for loop, in sequence, the loop terminate instantly and then the images load whenever they can load. You have zero control on that.

in my modified fiddle:
https://jsfiddle.net/0m5zxnw7/

this is the console output:
![image](https://user-images.githubusercontent.com/1194048/152124321-99b61a9e-6b22-48b5-bc04-b73a53f69f6e.png)

Now there are ways to solve this, and that is plain javascript unrelated to fabricJS.
You could use promises and use promise.all for example, or load the image one after the other.

FabricJS supporting promises will defiintely help
